### PR TITLE
feat(gas price): instruction precise gas price

### PR DESCRIPTION
**Pull request by olonho**
Test plan
---
Run per manual in emu-cost/README.md, meaningful costs appears.

### CODE REVIEW

File: runtime/runtime-params-estimator/emu-cost/large_regions.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// This Kotlin/Native utility takes /proc/<pid>/maps file to stdin and prints regions
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/emu-cost/large_regions.kt:1]: OK, I will remove it. It is utility used during debugging to understand which region  was the largest one, as /proc/<pid>/maps only shows addresses, not size.

File: runtime/runtime-params-estimator/emu-cost/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+RUN rustup toolchain uninstall nightly
+# Must match nearcore/rust-toolchain.
+RUN rustup default nightly-2020-03-19
+RUN rustup target add wasm32-unknown-unknown
+
+# install build dependencies for QEMU
+RUN apt-get update && apt-get install -y  g++ rsync zip openssh-server \
+     make apt-utils git sudo pkg-config libglib2.0-dev curl clang gdb
+
+# Optional: configure SSH
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/emu-cost/Dockerfile:17]: Unneeded, will remove.

File: runtime/runtime-params-estimator/emu-cost/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ Instrumentation of QEMU is implemented in the following manner. We install instrumentation callback which conditionally increments
+the instruction counter on every instruction during translation by QEMU's JIT, TCG. We activate counting when specific Linux syscall
+(currently, 0 aka sys_read) is executed with the certain arguments (file descriptor argument == 0xcafebabe or 0xcafebabf).
+On start event we clear instruction counter, on stop event we stop counting and return counted instructions into the buffer provided
+to read syscall. As result, NEAR benchmark will know the exact instruction counter passeed between two moments and this value
+is the pure function of Docker image used, Rust compiler version and the NEAR implementation and is fully reproducible.
+
+## Usage
+
+We build and run the cost estimator in the Docker container to make sure config is fully reproducible.
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/emu-cost/README.md:20]: Hmm,  interesting question, at least 4G is needed for benchmark to finish, not sure about how it affects results. Do not think it should.

File: runtime/runtime-params-estimator/src/cases.rs [hunk 4]
```diff
@@ -545,57 +518,66 @@ fn get_runtime_fees_config(measurement: &Measurements) -> RuntimeFeesConfig {
...
-        storage_iter_create_prefix_base: f64_to_gas(pure[&storage_iter_create_prefix_base]),
-        storage_iter_create_prefix_byte: f64_to_gas(pure[&storage_iter_create_prefix_byte]),
-        storage_iter_create_range_base: f64_to_gas(pure[&storage_iter_create_range_base]),
-        storage_iter_create_from_byte: f64_to_gas(pure[&storage_iter_create_from_byte]),
-        storage_iter_create_to_byte: f64_to_gas(pure[&storage_iter_create_to_byte]),
-        storage_iter_next_base: f64_to_gas(pure[&storage_iter_next_base]),
-        storage_iter_next_key_byte: f64_to_gas(pure[&storage_iter_next_key_byte]),
-        storage_iter_next_value_byte: f64_to_gas(pure[&storage_iter_next_value_byte]),
+        base: cost_as_gas(&pure, base),
+        read_memory_base: cost_as_gas(&pure, read_memory_base),
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/src/cases.rs:534]: Before it was very hard to understand which element was missing from the map, we were just seeing the misleading panic.

File: runtime/runtime-params-estimator/emu-cost/counter_plugin/test.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
+#include <stdint.h>
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/emu-cost/counter_plugin/test.c:1]: `make test`.

File: runtime/runtime-params-estimator/emu-cost/large_regions.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/emu-cost/large_regions.kt:1]: Regarding scaling insn to gas. I guess the whole point here is to make it gas price proportional to the actual computation request, not to the CPU horsepower, for example. So I believe that we can look at the table similar to
https://en.wikipedia.org/wiki/Instructions_per_second, and for example take  for reference Intel Core i7 Sandy Bridge (Intel Core i7 2600K in the table) which is said to execute 8.61 insn per clock tick on 3.4 Ghz clock frequency, which gives 
3.4*10^9 * 8.61 ~ 3 * 10^10 instructions per second, thus single instruction can be estimated as 3 * 10^4 of gas. However, current comparison of counted  insn vs. existing gas costs shows the following ratios:
 ``
base: time=126224222 insn=675705724 insn/time=5.35
read_memory_base: time=1629369577 insn=3051600275 insn/time=1.87
read_memory_byte: time=123816 insn=10130457 insn/time=81.82
write_memory_base: time=76445225 insn=762445430 insn/time=9.97
write_memory_byte: time=809907 insn=6999835 insn/time=8.64
read_register_base: time=639340699 insn=2810002290 insn/time=4.40
read_register_byte: time=63637 insn=255646 insn/time=4.02
write_register_base: time=0 insn=581238522 insn/time=n/a
write_register_byte: time=0 insn=104 insn/time=n/a
utf8_decoding_base: time=0 insn=1302414122 insn/time=n/a
utf8_decoding_byte: time=591904 insn=776744 insn/time=1.31
utf16_decoding_base: time=0 insn=2591064022 insn/time=n/a
utf16_decoding_byte: time=9095538 insn=52745880 insn/time=5.80
sha256_base: time=710092630 insn=5180024081 insn/time=7.29
sha256_byte: time=5536829 insn=53010127 insn/time=9.57
keccak256_base: time=710092630 insn=8747600581 insn/time=12.32
keccak256_byte: time=5536829 insn=45604690 insn/time=8.24
keccak512_base: time=710092630 insn=8565228839 insn/time=12.06
keccak512_byte: time=5536829 insn=86098618 insn/time=15.55
log_base: time=0 insn=0 insn/time=n/a
log_byte: time=0 insn=0 insn/time=n/a
storage_write_base: time=21058769282 insn=81774931785 insn/time=3.88
storage_write_key_byte: time=23447086 insn=155353981 insn/time=6.63
storage_write_value_byte: time=9437547 insn=55154630 insn/time=5.84
storage_write_evicted_byte: time=0 insn=0 insn/time=n/a
storage_read_base: time=19352220621 insn=49354605855 insn/time=2.55
storage_read_key_byte: time=4792496 insn=57489905 insn/time=12.00
storage_read_value_byte: time=139743 insn=373531 insn/time=2.67
storage_remove_base: time=109578968621 insn=60173708639 insn/time=0.55
storage_remove_key_byte: time=9512022 insn=74652152 insn/time=7.85
storage_remove_ret_value_byte: time=0 insn=591140 insn/time=n/a
storage_has_key_base: time=20019912030 insn=48063628422 insn/time=2.40
storage_has_key_byte: time=4647597 insn=57451705 insn/time=12.36
storage_iter_create_prefix_base: time=28443562030 insn=28443562030 insn/time=1.00
storage_iter_create_prefix_byte: time=442354 insn=442354 insn/time=1.00
storage_iter_create_range_base: time=25804628282 insn=25804628282 insn/time=1.00
storage_iter_create_from_byte: time=429608 insn=429608 insn/time=1.00
storage_iter_create_to_byte: time=1302886 insn=1302886 insn/time=1.00
storage_iter_next_base: time=24213271567 insn=24213271567 insn/time=1.00
storage_iter_next_key_byte: time=0 insn=0 insn/time=n/a
storage_iter_next_value_byte: time=1343211668 insn=1343211668 insn/time=1.00
touching_trie_node: time=1 insn=1 insn/time=1.00
promise_and_base: time=0 insn=0 insn/time=n/a
promise_and_per_promise: time=672136 insn=4304689 insn/time=6.40
promise_return: time=34854215 insn=497252118 insn/time=14.27
``
Which hints that memory intensive operations (probably due to cache amortization) are 10^2 more CPU intensive compared to raw timing than it seemed. 

What we could do, is try to normalize cost of some CPU-intensive operation (sha256 looks good one), and use the factor we see here as insn cost, i.e. smth line 7-8.

File: runtime/runtime-params-estimator/src/testbed_runners.rs [hunk 6]
```diff
@@ -8,7 +8,6 @@ use near_primitives::transaction::{Action, SignedTransaction};
 use rand::Rng;
```
[the remark this answers is not in the document]
Comment (olonho) [reply runtime/runtime-params-estimator/src/testbed_runners.rs:8]: Explained in https://nearhack3r.blogspot.com/2020/04/precise-gas-metering.html

File: runtime/runtime-params-estimator/emu-cost/large_regions.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nearmax) [reply runtime/runtime-params-estimator/emu-cost/large_regions.kt]: It is a good idea to use sha256 for normalization.

**Pull request merged**